_SPIKE_.md

```diff
@@ -0,0 +1,244 @@
+>   (CRUD, KeyValue, etc.) and what datasource to use under the hood.
+> - `@loopback/boot` processes this configuration and registers appropriate
+>   repositories & controllers with the app.
+
+In this Spike, I am demonstrating a PoC implementation of an extensible booter
+that processed model configuration files in JSON formats and uses 3rd-party
+plugins to build repository & controller classes at runtime.
+
+## Basic use
+
+Create `src/public-models` directory in your project. For each model you want to
+expose via REST API, add a new `.config.ts` file that's exporting the model
+configuration.
+
+Example (`src/public-models/product.config.ts`):
+
+```ts
+import {ModelApiConfig} from '@loopback/rest-crud';
+
+module.exports = <ModelApiConfig>{
+  model: 'Product',
+  pattern: 'CrudRest',
+  dataSource: 'db',
+  basePath: '/products',
+};
+```
+
+## Implementation
+
+The solution has the following high-level parts:
+
+1. A new package `@loopback/model-api-builder` defines the contract for plugins
+   (extensions) contributing repository & controller builders.
+
+2. A new booter `RestBooter` that loads all JSON files from
+   `/public-models/{model-name}.config.json`, resolves model name into a model
+   class (via Application context), finds model api builder using
+   Extension/ExtensionPoint pattern and delegates the remaining work to the
+   plugin.
+
+3. An official model-api-builder plugin for building CRUD REST APIs using
+   `DefaultCrudRepository` implementation. The plugin is implemented inside the
+   recently introduced package `@loopback/rest-crud`.
+
+Under the hood, the CRUD REST API builder performs two steps:
+
+1. It defines a model-specific repository class, e.g. `ProductRepository` and
+   registers it with the Application context for Dependency Injection. In the
+   future, we can make this step optional and allow app developers to supply
+   their own repository class (e.g. as created by `lb4 repository`) 💪
+
+   The repository class is decorated to receive the dataSource instance via DI.
+
+2. Then a model-specific controller class is defined, e.g. `ProductController`
+   and decorated to receive the repository instance via DI. The controller is
+   registered with the Application context to get it included in the public API.
+
+I feel it's important to use model-named classes (e.g. `ProductController`)
+instead of reusing the same controller/repository class name for all models.
+Model-named classes make error troubleshooting and debugging easier, because
+stack traces include model name in function names. (Compare
+`ProductController.replaceById` with `CrudRestControllerImpl.replaceById` -
+which one is more useful?)
+
+In my proposal, model-config files are in JSON format to make programmatic edits
+easier. This has a downside in TypeScript projects - these config files must
+live outside `src` because TypeScript does not copy arbitrary JSON files.
```

Comment:
IIRC, we can configure the tsconfig to allow json files inside `src`.

Reply:
Sure. However, AFAIK, vanilla `tsc` copies to `dist` only those JSON files that are directly `import`ed from a `.ts` file. Model config files are not imported, they are processed at runtime.

Anyhow, I'd like to start with `.ts` files for the initial version. We can add support for other flavors like `JSON` files later, I believe such improvement should be a small & backwards compatible change.